### PR TITLE
Update launch condition column

### PR DESCRIPTION
**Pull request by wild-endeavor**
## Tracking issue
This introduces a new way to specify schedules but is not yet complete.

### IDL
* Add a trigger enum to the execution metadata.
* Remove the index from artifact binding

### DB
Add a new migration to add a `launch_condition_type` column to the launch plan database.  This column is searchable.

## Why are the changes needed?
Cleaner way to specify schedules in the future.

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] All commits are signed-off.

### CODE REVIEW

File: flyteadmin/pkg/repositories/config/migrations.go [hunk 1]
```diff
@@ -1182,7 +1182,55 @@ var NoopMigrations = []*gormigrate.Migration{
...
+				Spec       []byte     `gorm:"not null"`
+				WorkflowID uint       `gorm:"index"`
+				Closure    []byte     `gorm:"not null"`
+				// GORM doesn't save the zero value for ints, so we use a pointer for the State field
+				State *int32 `gorm:"default:0"`
+				// Hash of the launch plan
+				Digest       []byte
+				ScheduleType LaunchPlanScheduleType
+				// store the type of event that this launch plan is triggered by, can be empty, or SCHED
+				LaunchConditionType string `gorm:"null"`
```
Comment (wild-endeavor) [inline flyteadmin/pkg/repositories/config/migrations.go:1209]: *

File: flyteidl/protos/flyteidl/core/artifact_id.proto [hunk 2]
```diff
@@ -17,8 +17,6 @@ message ArtifactKey {
 
 // Only valid for triggers
 message ArtifactBindingData {
-  uint32 index = 1;
```
Comment (katrogan) [inline flyteidl/protos/flyteidl/core/artifact_id.proto:20]: why was this removed? does this wire number need to be marked reserved?

File: flyteadmin/pkg/repositories/models/launch_plan.go [hunk 3]
```diff
@@ -19,7 +19,14 @@ const (
 	LaunchPlanScheduleTypeRATE LaunchPlanScheduleType = "RATE"
 )
 
-// Database model to encapsulate a launch plan.
+type LaunchConditionType string
+
+const (
+	// LaunchConditionTypeSCHED is the const representing the launch plan has a CRON type of schedule
```
Comment (katrogan) [inline flyteadmin/pkg/repositories/models/launch_plan.go:25]:
```suggestion
	// LaunchConditionTypeSCHED is the const representing the launch plan has a trigger type of schedule
```

File: flyteadmin/pkg/repositories/models/launch_plan.go [hunk 4]
```diff
@@ -29,8 +36,9 @@ type LaunchPlan struct {
 	// GORM doesn't save the zero value for ints, so we use a pointer for the State field
 	State *int32 `gorm:"default:0"`
 	// Hash of the launch plan
-	Digest       []byte
-	ScheduleType LaunchPlanScheduleType
+	Digest              []byte
+	ScheduleType        LaunchPlanScheduleType
+	LaunchConditionType *LaunchConditionType
```
Comment (katrogan) [inline flyteadmin/pkg/repositories/models/launch_plan.go:41]: just curious, should this be indexed?

**Review by katrogan — COMMENTED** [review]
LGTM

File: flyteidl/protos/flyteidl/core/artifact_id.proto (hunk 2 above)
Comment (wild-endeavor) [reply flyteidl/protos/flyteidl/core/artifact_id.proto:20]: i don't want it anymore.

File: flyteadmin/pkg/repositories/models/launch_plan.go (hunk 4 above)
Comment (wild-endeavor) [reply flyteadmin/pkg/repositories/models/launch_plan.go:41]: yeah i added a conditional index in the migration - https://github.com/flyteorg/flyte/pull/4903/files#diff-7189b1a0f8ed581ca15085037de7cd70b60305b90e42cc60f7eda13796577788R1221.  is that okay?

Comment (katrogan) [reply flyteadmin/pkg/repositories/models/launch_plan.go:41]: maybe add here for consistency?

**Review by katrogan — APPROVED** [review]

File: flyteadmin/pkg/repositories/models/launch_plan.go (hunk 4 above)
Comment (wild-endeavor) [reply flyteadmin/pkg/repositories/models/launch_plan.go:41]: let me try...

Comment (wild-endeavor) [reply flyteadmin/pkg/repositories/models/launch_plan.go]: k - i think that is it https://github.com/flyteorg/flyte/pull/4903/files#diff-34110b943c015fdb7456f2c8fa9759e25ff476bf32ed261b2945f7bb928f56dcR41.

but i kinda want to keep the explicit index in the migration if that's okay?

Comment (wild-endeavor) [reply flyteadmin/pkg/repositories/models/launch_plan.go]: actually, nm, it's cleaner to just use gorm, as much as i dislike it.  I added a type, dropped the column in my local sandbox postgres and re-ran the migration with the new annotation.  Gorm ran these two queries.
```
ALTER TABLE "launch_plans" ADD "launch_condition_type" varchar(32)
```

and

```
 CREATE INDEX IF NOT EXISTS "idx_launch_plans_launch_conditions" ON "launch_plans" ("launch_condition_type") WHERE launch_condition_type is not null
```

File: flyteadmin/pkg/repositories/config/migrations.go [hunk 5]
```diff
@@ -1182,7 +1182,43 @@ var NoopMigrations = []*gormigrate.Migration{
 	},
 }
 
-var Migrations = append(LegacyMigrations, NoopMigrations...)
+var ContinuedMigrations = []*gormigrate.Migration{
```
Comment (katrogan) [inline flyteadmin/pkg/repositories/config/migrations.go:1185]: can you add a comment here explaining what ContinuedMigrations is and that we should now add new migrations here?

**Review by katrogan — DISMISSED** [review]

**Review by katrogan — APPROVED** [review]

**Review by squiishyy — APPROVED** [review]
